test/hotspot/jtreg/compiler/lib/template_framework/CodeFrame.java

```diff
@@ -0,0 +1,156 @@
+ * questions.
+ */
+
+package compiler.lib.template_framework;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * The {@link CodeFrame} represents a frame (i.e. scope) of code, appending {@link Code} to the {@code 'codeList'}
+ * as {@link Token}s are rendered, and adding names to the {@link NameSet}s with {@link Template#addName}.
+ * {@link Hook}s can be added to a frame, which allows code to be inserted at that location later.
+ * When a {@link Hook} is {@link Hook#set}, this separates the Template into an outer and inner
+ * {@link CodeFrame}, ensuring that names that are {@link Template#addName}'d inside the inner frame
+ * are only available inside that frame.
+ *
+ * <p>
+ * On the other hand, each {@link TemplateFrame} represents the frame (or scope) of exactly one use of a
+ * Template.
+ *
+ * <p>
+ * For simple Template nesting, the {@link CodeFrame}s and {@link TemplateFrame}s overlap exactly.
+ * However, when using {@link Hook#insert}, we simply nest {@link TemplateFrame}s, going further "in",
+ * but we jump to an outer {@link CodeFrame}, ensuring that we insert {@link Code} at the outer frame,
+ * and operating on the names of the outer frame. Once the {@link Hook#insert}ion is complete, we jump
+ * back to the caller {@link TemplateFrame} and {@link CodeFrame}.
+ */
+class CodeFrame {
+    public final CodeFrame parent;
+    private final List<Code> codeList = new ArrayList<Code>();
+    private final Map<Hook, Code.CodeList> hookCodeLists = new HashMap<>();
+
+    /**
+     * The {@link NameSet} is used for variable and fields etc.
+     */
+    final NameSet names;
+
+    private CodeFrame(CodeFrame parent, boolean isTransparentForNames) {
+        this.parent = parent;
+        if (parent == null) {
+            // NameSet without any parent.
+            this.names = new NameSet(null);
+        } else if (isTransparentForNames) {
+            // We use the same NameSet as the parent - makes it transparent.
+            this.names     = parent.names;
+        } else {
+            // New NameSet, to make sure we have a nested scope for the names.
+            this.names     = new NameSet(parent.names);
+        }
+    }
+
+    /**
+     * Creates a base frame, which has no {@link #parent}.
+     */
+    public static CodeFrame makeBase() {
+        return new CodeFrame(null, false);
+    }
+
+    /**
+     * Creates a normal frame, which has a {@link #parent} and which defines an inner
+     * {@link NameSet}, for the names that are generated inside this frame. Once this
+     * frame is exited, the name from inside this frame are not available anymore.
+     */
+    public static CodeFrame make(CodeFrame parent) {
+        return new CodeFrame(parent, false);
+    }
+
+    /**
+     * Creates a special frame, which has a {@link #parent} but uses the {@link NameSet}
+     * from the parent frame, allowing {@link Template#defineName} to persist in the outer
```

Comment:
`defineName` -> `addName`?

Reply:
Good catch! Left over from a previous refactoring. `javadoc` does not complain about it, because it seems it does not look at anything that is not `public` :see_no_evil: